### PR TITLE
[WIP] Classifier ES6 rewrite

### PR DESCRIPTION
**Pull request by CKrawczyk**
This PR rewrites the classifier interface into ES6 and add tests for each component.  Here is a list of what has been done and what still needs to be done:
- [x] Convert `tutorial-button`
- [x] Convert `mini-course-button`
    - [x] extract common elements of `tutorial-button` and `mini-course-button` to higher order component (`restart-button`)
    - [x] write tests for `tutorial-button` and `mini-course-button`
- [x] Convert `warning-banner`
    - [x] write tests for `warning-banner`
- [x] Convert `classification-summary`
    - [x] write tests for `classification-summary`
    - [x] figure out how to get tests that require `.cjsx` to run (I think we will need `coffee:coffee-react/register`, but that still did not work)
- [x] Convert `world-wide-telescope` ([PR](https://github.com/zooniverse/Panoptes-Front-End/pull/3463) is up)
    - [x] write tests for `world-wide-telescope`
- [x] Convert `frame-annotator`
    - [x] write tests for `frame-annotator`
- [x] Convert `metadata-based-feedback` (Converted and tested, but unmerged since functionality changes are needed for Space Warps use)
    - [x] write tests for `metadata-based-feedback`
- [ ] Convert `index`
    - [x] Split out independent components to different files (`render-summary`, `render-expert-options`, `classifier-helpers`, `index`, and `classifier`).
    - [ ] write tests for all these components

For the most part the classifier works, but I have yet to test any project using the gravity-spy or world-wide-telescope flags or subjects with expert feedback.

One bug that still needs to be addressed is the `gold standard` checkbox of the `render-expert-options` popover is not causing a re-render of the component.  This only happens when it is clicked from the `classification-summary` page, the same exact component works just fine on the `classification` page.

As there are many components that are displayed only if particular flags are set, I have written a `VisibilityWrapper` that takes the prop `visible` and displays its children only if it is truthy.  This avoids the need to place a bunch of `if` statements in the `jsx`.

Other things that would be nice, but might be better for a later PR:
- [ ] Convert and write tests for the `task` folder
- [ ] Convert and write tests for the `drawing-tools` folder

I will be on vacation for the next month, if anyone wants to help with this feel free to commit directly to the branch.

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [x] Did you deploy a staging branch? https://classifier-rewrite.pfe-preview.zooniverse.org/
- [ ] Did you get any type checking errors from [Babel Typecheck](https://github.com/codemix/babel-plugin-typecheck)

## Optional

- [ ] If it's a new component, is it in ES6? Is it clear of warnings from ESLint?
- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
@CKrawczyk I've rebased this against the latest master.

**Comment by eatyourgreens** [comment]
Woo! Tests are passing again (but only because I skipped the `ClassificationSummary` tests, which crash with an `illegal import declaration` error.)

**Comment by CKrawczyk** [comment]
Yeah, until we figure out how to get `.cjsx` files to import into tests (or rewrite everything into ES6) that test will fail.

**Comment by eatyourgreens** [comment]
`coffee-react` should fix `.cjsx` files. The failure seems to be in one of the tasks.
```
at Object.<anonymous> (/Users/odonnell/zooniverse/Panoptes-Front-End/app/classifier/tasks/single.cjsx:4:21)
  at Object.<anonymous> (/Users/odonnell/zooniverse/Panoptes-Front-End/app/classifier/tasks/single.cjsx:1:1)
```

**Comment by CKrawczyk** [comment]
Oh, good to know.  Since I was not able to get the test to even run before, I never got around to seeing if the test even worked in the first place.  That file was more of a place holder than anything else.

I am new to writing test code, so there are likely errors in the way I wrote the test.

**Comment by eatyourgreens** [comment]
I don't think there's anything wrong with your tests. The spec file fails on `import ClassificationSummary` so the tests don't even run.

**Comment by eatyourgreens** [comment]
Looking further into this, the components that crash the tests are components that import from `markdownz`. Switching from `require` to `import` doesn't seem to help either.

**Comment by eatyourgreens** [comment]
@CKrawczyk testing coffeescript components is a seperate issue. Let's move that conversation over to #3417

**Comment by eatyourgreens** [comment]
Tests are passing now 👍 but there's been some recent work on the classifier coffeescript, so maybe some fiddly merging to copy those changes across to the new ES6 file.

**Comment by srallen** [comment]
I've rebased and finished the rewrite of the `MetadataBasedFeedback` component in my own branch: https://github.com/zooniverse/Panoptes-Front-End/tree/space-warps-feedback-refactor 

I could open a PR to merge into this branch, but PR is already getting very large. Not sure how we want to approach this.

**Comment by eatyourgreens** [comment]
@srallen for #3402, I cherry-picked changes from this branch into my PR, then rebased this PR against master after #3402 was merged. Would that work for the metadata feedback too?

**Comment by eatyourgreens** [comment]
@wgranger Notifications tests should be fixed now. See e0941b3d83793ea830385f4536cbe4285568c6bb

**Comment by CKrawczyk** [comment]
@srallen It might be best if you submit your PR separately if you want the changes sooner rather than later.  I am a bit busy with the Portsmouth end of my job for the rest of this week, so I will not be able to start working on this branch again until next week.

That being said, I can always split this branch into a bunch of smaller PRs if we want to keep the size down.

**Comment by srallen** [comment]
Separate PR sounds good. I can take a look at writing a few tests for it and then submit one. Not sure when I'll get back to that this week. Probably Thursday or Friday.

**Comment by eatyourgreens** [comment]
I've moved the changes to get CJSX tests working into #3428 Once that's merged, we can rebase this branch against master.

**Comment by eatyourgreens** [comment]
Rebased to merge in @wgranger's recent changes to the subject viewer, and staged at https://classifier-rewrite.pfe-preview.zooniverse.org/

**Comment by srallen** [comment]
We may want to consider pulling the classification summary rewrite and tests out into its own PR and  incorporate the split added in #3434.

**Comment by eatyourgreens** [comment]
I've opened #3440 for testing the `tasks` directory.

**Comment by srallen** [comment]
Just an update on the metadata based feedback component: While writing tests, I found a bug, so I need to resolve that before making a PR.

**Comment by eatyourgreens** [comment]
#3439 probably conflicts with this PR too.

**Comment by srallen** [comment]
Ok, so I have more info from the SW team that the metadata feedback is going to need some more work, so the rest of the classifier rewrite shouldn't wait for the metadata feedback's component rewrite. (see: https://github.com/zooniverse/Panoptes-Front-End/issues/3361#issuecomment-276326472)

**Comment by srallen** [comment]
I can take a look at writing tests for the tutorial, mini-course, restart buttons. Also will push up as separate branch and PR.

**Comment by CKrawczyk** [comment]
I will start work on resolving the conflicts with the classification-summary and push it up as a separate branch and PR.

**Comment by CKrawczyk** [comment]
I just split off the `warning-banner` component and text to a new branch and PR (#3453).

**Comment by eatyourgreens** [comment]
@CKrawczyk those PRs are merged now so you should be able to remove the corresponding commits from this branch.

**Comment by srallen** [comment]
Tests for restart button is in #3455

**Comment by CKrawczyk** [comment]
I just re-based onto the current master.  This has not added changes from #3448 into the conversion of `index`.

File: app/classifier/classification-summary.jsx [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import React from 'react';
 import tasks from './tasks/index.coffee';
 import { TextSplit } from 'seven-ten';
+import tasks from './tasks';
```
Comment (eatyourgreens) [inline app/classifier/classification-summary.jsx]: There's a duplicate import here.

File: app/classifier/classification-summary.jsx [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import React from 'react';
 import tasks from './tasks/index.coffee';
```
Comment (eatyourgreens) [inline app/classifier/classification-summary.jsx]: I think I got rid of this file.

**Comment by srallen** [comment]
Working on frame-annotator tests in separate branch.

File: app/classifier/classification-summary.jsx (hunk 1 above)
Comment (CKrawczyk) [inline app/classifier/classification-summary.jsx]: Good catch, there were a few other import errors that were causing the tests to fail.  I did not catch all of them in the rebase.  I will fix this up so at least the tests pass.

**Comment by srallen** [comment]
It should be noted that the `subject-viewer` and `frame-viewer` and associated subject type viewers are in the components folder and were missed in the checklist above. I'd be for moving these into the classifier folder to make it clear they are tightly integrated into the classifier even though the viewers are also used by talk.

**Comment by srallen** [comment]
Added a few test for the frame annotator in #3482

**Comment by srallen** [comment]
Closing since most of this is done and will open a separate issue to track what is left.

**Pull request closed without merging**